lib/MongoDB/ClientSession.pm

```diff
@@ -215,6 +259,242 @@ sub advance_operation_time {
+  transaction.
+
+=cut
+
+sub start_transaction {
+    my ( $self, $opts ) = @_;
+
+    MongoDB::TransactionError->throw("Transaction already in progress")
+        if $self->_in_transaction_state( 'starting', 'in_progress' );
+
+    MongoDB::ConfigurationError->throw("Transactions are unsupported on this deployment")
+        unless $self->client->_topology->_supports_transactions;
+
+    $opts ||= {};
+    $opts = { %{ $self->options->{defaultTransactionOptions} }, %$opts };
+
+    $self->_set__current_transaction_settings( $opts );
+
+    # Trigger build of write_concern - must error on start of transaction
+    $self->_clear_transaction_write_concern;
+    $self->_transaction_write_concern;
+
+    $self->_set__transaction_state('starting');
+
+    $self->_increment_transaction_id;
+
+    $self->_set__active_transaction( 1 );
+    $self->_set__has_transaction_operations( 0 );
+
+    return;
+}
+
+sub _increment_transaction_id {
+    my $self = shift;
+    return if $self->_active_transaction;
+
+    $self->_server_session->transaction_id->binc();
+}
+
+=method commit_transaction
+
+Commit the current transaction. This will use the writeConcern set on this transaction.
+
+=cut
+
+sub commit_transaction {
+    my $self = shift;
+
+    MongoDB::TransactionError->throw("No transaction started")
+        if $self->_transaction_state eq 'none';
+
+    # Error message tweaked to use our function names
+    MongoDB::TransactionError->throw("Cannot call commit_transaction after calling abort_transaction")
+        if $self->_transaction_state eq 'aborted';
+
+    # Commit can be called multiple times - even if the transaction completes
+    # correctly. Setting this here makes sure we dont increment transaction id
+    # until after another command has been called using this session
+    $self->_set__active_transaction( 1 );
+
+    eval {
+        $self->_send_end_transaction_command( 'committed', [ commitTransaction => 1 ] );
+    };
+    if ( my $err = $@ ) {
+        # catch and re-throw after retryable errors
+        # TODO maybe need better checking logic that theres actually an error code in output?
+        my $err_code;
+        if ( $err->can('result') ) {
+            if ( $err->result->can('output') ) {
+                $err_code = $err->result->output->{codeName};
+                $err_code ||= $err->result->output->{writeConcernError}
+                  ? $err->result->output->{writeConcernError}->{codeName}
+                  : ''; # Empty string just in case
+            }
+        }
+        # If its a write concern error, retrying a commit would still error
+        unless ( defined( $err_code ) && grep { $_ eq $err_code } qw/
+            CannotSatisfyWriteConcern
+            UnsatisfiableWriteConcern
+            UnknownReplWriteConcern
+            NoSuchTransaction
+        / ) {
+            push @{ $err->error_labels }, 'UnknownTransactionCommitResult';
```

Comment:
This label also needs to be applied to server selection errors, network errors, network timeout errors, and retryable write errors.

Reply:
Yep, those are the only occurances for it not being applied - in the latest version have had to put int he actual error code numbers in as some of the spec tests dont actually set the codeName... (`error-labels.json` in the last two omit tests)